### PR TITLE
Add a FAQ on contention

### PR DESCRIPTION
**Pull request by knz**
I've been seeing this _so_ many times everywhere (on the forum, on gitter, from sales etc) that I am surprised we didn't FAQ this earlier. Here it is.

### CODE REVIEW

**Comment by knz** [comment]
Once the copy-editing is done I'll do the propagation to the other version directories

File: v2.1/sql-faqs.md [hunk 1]
```diff
@@ -39,6 +39,71 @@ For example, this is how you’d use `RETURNING` to return an auto-generated [`S
...
+together:
+
+- There are multiple concurrent transactions or statements (sent by
+  multiple clients connected simultaneously to a single CockroachDB
+  cluster).
+- They operate on the same data, specifically
+  over table rows with the same index key values (either on [primary
+  keys](primary-key.html) or secondary [indexes](indexes.html)) or using
+  index key values that are close to each other, and thus place the
+  indexed data on the same [data ranges](architecture/overview.html).
```
Comment (benesch) [inline v2.1/sql-faqs.md]: table rows with the same index key values _within the same column family_, no?a

**Comment by knz** [comment]
The title must stay because *we* often talk about contention in our explanations to users (and in blog posts, and other things) and it would be good for people to have a definition/explanation to search for if they want to understand our answers better.

However I agree we must provide additional entry points to make this information better discoverable. I'll make some attempt. Will let you know.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3156#-:-LCjDPQA94zPd5JkSYh9:b90rwug)*

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3156#-:-LD-JJZV9veaTwJx9MX9:bnfp4nl)*

File: v2.1/performance-best-practices-overview.md [hunk 2]
```diff
@@ -226,3 +226,88 @@ This query returns the account balances of the customers.
 If you have long-running queries (such as analytics queries that perform full table scans) that can tolerate slightly out-of-date reads, consider using the [`... AS OF SYSTEM TIME` clause](select-clause.html#select-historical-data-time-travel). Using this, your query returns data as it appeared at a distinct point in the past and will not cause [conflicts](architecture/transaction-layer.html#transaction-conflicts) with other concurrent transactions, which can increase your application's performance.
 
 However, because `AS OF SYSTEM TIME` returns historical data, your reads might be stale.
+
+## Understanding and Avoiding Transaction Contention
+
+Transaction contention occurs when the following three conditions hold
```
Comment (jseldess) [inline v2.1/performance-best-practices-overview.md]: nit: `hold together` > `are met`.

File: v2.1/performance-best-practices-overview.md [hunk 3]
```diff
@@ -226,3 +226,88 @@ This query returns the account balances of the customers.
...
+
+Transaction contention can also increase the rate of transaction
+restarts, and thus make the proper implementation of [client-side
+transaction
+retries](transactions.html#client-side-transaction-retries) more
+critical.
+
+To avoid contention, multiple strategies can be applied:
+
+- Use index key values with a more random distribution of values, so
```
Comment (jseldess) [inline v2.1/performance-best-practices-overview.md]: Probably best to link directly to the specific FAQ you have in mind.

File: v2.1/performance-best-practices-overview.md [hunk 4]
```diff
@@ -226,3 +226,88 @@ This query returns the account balances of the customers.
...
+  expressions](common-table-expressions.html) to group multiple
+  [`SELECT`](select-clause.html) and
+  [`INSERT`](insert.html)/[`UPDATE`](update.html)/[`DELETE`](delete.html)/[`UPSERT`](upsert.html)
+  clauses together in a single SQL statement.
+
+- When replacing values in a row, use [`UPSERT`](upsert.html) and
+  specify values for all columns in the inserted rows. This will
+  usually have the best performance under contention, compared to
+  combinations of [`SELECT`](select-clause.html),
+  [`INSERT`](insert.html) and [`UPDATE`](update.html).
```
Comment (jseldess) [inline v2.1/performance-best-practices-overview.md]: nit: place comma before the `and` (the "oxford comma").

File: v2.1/performance-best-practices-overview.md [hunk 5]
```diff
@@ -226,3 +226,88 @@ This query returns the account balances of the customers.
...
+
+- If the application strictly requires operating on very few different
+  index key values, consider using [`ALTER ... SPLIT
+  AT`](split-at.html) so that each index key value can be served by
+  a separate group of nodes in the cluster.
+
+It is always best to avoid contention as much as possible via the
+design of the schema and application. However, sometimes contention is
+unavoidable. To maximize performance in the presence of contention,
+you'll need to maximize the performance of a single range.
```
Comment (jseldess) [inline v2.1/performance-best-practices-overview.md]: Is this paragraph introducing the following bulleted list? If so, use a semicolon instead of a period at the end of the last sentence.

File: v2.1/sql-faqs.md [hunk 6]
```diff
@@ -39,6 +39,30 @@ For example, this is how you’d use `RETURNING` to return an auto-generated [`S
 > INSERT INTO users (name) VALUES ('mike') RETURNING id;
 ~~~
 
+## What is transaction contention?
+
+Transaction contention occurs when transactions issued from multiple
+clients at the same time operate on the same data.
+
+This can cause transactions to wait on each other and decrease
```
Comment (jseldess) [inline v2.1/sql-faqs.md]: This doesn't need to be its own paragraph. Please make it the second sentence of the previous paragraph.

File: v2.1/sql-faqs.md [hunk 7]
```diff
@@ -39,6 +39,30 @@ For example, this is how you’d use `RETURNING` to return an auto-generated [`S
...
+performance, like when many people try to check out with the same
+cashier at a store.
+
+For more information about contention, see [Understanding and Avoiding
+Transaction
+Contention](performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention).
+
+## Why can I not get more operations per second by increasing the number of nodes?
```
Comment (jseldess) [inline v2.1/sql-faqs.md]: I'd rephrase this a bit: `Why would increasing the number of nodes not result in more operations per second?`

**Comment by knz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[v2.1/performance-best-practices-overview.md, line 232 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LCjL-7hC8aUPf-Mgdeh-r3-232:-LD7ApS75Dacb0jgOJl3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/performance-best-practices-overview.md#L232)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nit: `hold together` > `are met`.
</blockquote></details>

Done.

---

*[v2.1/performance-best-practices-overview.md, line 270 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LCjL-7hC8aUPf-Mgdeh-r3-270:-LD7BH7SBhYtUDd3LLIb:b-4dhtgr) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/performance-best-practices-overview.md#L270)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Probably best to link directly to the specific FAQ you have in mind.
</blockquote></details>

There are multiple. How do you suggest to do that?

---

*[v2.1/performance-best-practices-overview.md, line 287 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LCjL-7hC8aUPf-Mgdeh-r3-287:-LD7At197kopJiw6Lwtv:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/performance-best-practices-overview.md#L287)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nit: place comma before the `and` (the "oxford comma").
</blockquote></details>

Done.

---

*[v2.1/performance-best-practices-overview.md, line 305 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LCjL-7hC8aUPf-Mgdeh-r3-305:-LD7B0Gk4OGaU4xforZy:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/performance-best-practices-overview.md#L305)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Is this paragraph introducing the following bulleted list? If so, use a semicolon instead of a period at the end of the last sentence.
</blockquote></details>

Done.

---

*[v2.1/sql-faqs.md, line 47 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LChse2PJgFjb1vy23lJ-r3-47:-LD7B5sj0bcr8PsZJc8N:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/sql-faqs.md#L47)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

This doesn't need to be its own paragraph. Please make it the second sentence of the previous paragraph.
</blockquote></details>

Done.

---

*[v2.1/sql-faqs.md, line 55 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LChse2PJgFjb1vy23lJ-r3-55:-LD7B9vDAikYWYnawZVM:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/sql-faqs.md#L55)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

I'd rephrase this a bit: `Why would increasing the number of nodes not result in more operations per second?`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3156)*

**Comment by jseldess** [comment]
*[v2.1/performance-best-practices-overview.md, line 270 at r3](https://reviewable.io/reviews/cockroachdb/docs/3156#-LCjL-7hC8aUPf-Mgdeh-r3-270:-LD7NWQICGyegQU6xe8G:baaqp0h) ([raw file](https://github.com/cockroachdb/docs/blob/1f112e524e3365c3dc5e67c780c93ac22b1f91b8/v2.1/performance-best-practices-overview.md#L270)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

There are multiple. How do you suggest to do that?
</blockquote></details>

You're right. You can just leave it as-is. Users will see the toc and jump to the FAQ that interests them.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3156)*
